### ISSUE TITLE
Unstable error "Unable to establish one or more of the specified browser connections" after Chrome update (v83)

### ISSUE THREAD

**Issue by DirDash** [DevExpress/testcafe#5212]
### What is the Test Scenario?
TestCafe tests in TeamCity against headless Chrome in parallel.
 
### What is the Current behavior?
After Chrome update to v83 TestCafe tests periodically (in about half the times) do not start with the following error:
```
GeneralError: Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure.
      at BrowserSet._waitConnectionsOpened (E:\BuildAgent\work\6d1660a0b0f4fce5\testcafe\node_modules\testcafe\src\runner\browser-set.ts:91:30)
      at E:\BuildAgent\work\6d1660a0b0f4fce5\testcafe\node_modules\testcafe\src\runner\browser-set.ts:114:35
      at processTicksAndRejections (internal/process/task_queues.js:94:5)
      at Bootstrapper._getBrowserConnections (E:\BuildAgent\work\6d1660a0b0f4fce5\testcafe\node_modules\testcafe\src\runner\bootstrapper.ts:215:16)
      at async Promise.all (index 0)
      at Bootstrapper._bootstrapParallel (E:\BuildAgent\work\6d1660a0b0f4fce5\testcafe\node_modules\testcafe\src\runner\bootstrapper.ts:391:38)
      at Bootstrapper.createRunnableConfiguration (E:\BuildAgent\work\6d1660a0b0f4fce5\testcafe\node_modules\testcafe\src\runner\bootstrapper.ts:424:42) {
    code: 'E1004',
    data: []
  }
  npm ERR! code ELIFECYCLE
  npm ERR! errno 1
  npm ERR! test.testcafe@1.0.0 test:teamcity
  npm ERR! Exit status 1
  npm ERR! 
  npm ERR! Failed at the se.test.testcafe@1.0.0 test:teamcity script.
  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
  npm ERR! A complete log of this run can be found in:
  npm ERR!     C:\Users\TC_BuildService\AppData\Roaming\npm-cache\_logs\2020-06-16T13_04_02_492Z-debug.log
  Process exited with code 1
  Process exited with code 1 (Step: Tests (Command Line))
  Tests (Command Line) failed
```

Before the update these errors were very very rare.

It happens not only in TeamCity, but on local run too (less often).

Using `chrome:headless:userProfile` or/and `chrome:headless --no-sandbox` didn't help.

### What is the web application and TestCafe test code?

Parameters:
* target browser: **chrome:headless**
* concurrency level: **6**
* hostname: **localhost**
* port1: **1337**
* port2: **1338**
* skipJsErrors: **false**
* skipUncaughtErrors: **true**

<details>
<summary>TestCafe runner code:</summary>

```js
const createTestCafe = require('testcafe');
const path = require('path');
const config = require('./.testcafe.config');

let testcafe = null;
createTestCafe(config.hostname, config.port1, config.port2)
    .then(tc => {
        testcafe = tc;

        const runner = testcafe
            .createRunner()
            .browsers(config.browsers)
            .src(config.src)
            .concurrency(config.concurrency)
            .reporter(config.reporter);

        return runner.run({
            ...config.runnerOptions,
            quarantineMode: true
        });
    })
    .then(function(failedCount) {
        testcafe.close();
        process.exit(failedCount ? 1 : 0);
    })
    .catch(function(error) {
        console.error(error);
        testcafe.close();
        process.exit(1);
    });
```
</details>

<details>
<summary>Custom configuration file:</summary>

```js
{
    hostname: 'localhost',
    port1: 1337,
    port2: 1338,
    browser: 'chrome:headless',
    src: './tests/*.js',
    concurrency: 6,
    reporter: 'teamcity',

    screenshots: {
        fullPage: false,
        takeOnFails: false
    },

    runnerOptions: {
        skipJsErrors: false,
        skipUncaughtErrors: true,
        pageLoadTimeout: 15000,
        selectorTimeout: 6000,
        assertionTimeout: 6000
    }
}
```
</details>
 
### Environment details:
 
* testcafe version: 1.8.6
* node.js version: 12.14.1
* browser name and version: Chrome 83.0.4103.106 / Windows 10
* platform and version: Microsoft Windows Server 2016 Standard
* TeamCity: 2019.2.4 (build 72059)
* testcafe-reporter-teamcity: 1.0.10

### Comments

I'm sorry I can't provide a public link or a stable repro.

Please tell me which parameters I should pay attention at and which test configuration I should try. Could it be really related to last Chrome update?

If you need additional info I'm happy to provide it.

**Comment by doverradio** [comment]
Hello, I am also facing this same issue.  Mysterious `Unable to establish one or more of the specified browser connections.  This can be caused by network issues or remote device failure.`  

and got error code:
`code: 'E1004',`

**Comment by fissshka** [comment]
Hi. Same for me. Constantly face the issue with `Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure` error. 
Please, help to find out the workarounds.

**Comment by AndreyBelym** [comment]
Thank you for your reports. Please share an example project that can be used to reproduce the issue and specify your OS, browser and TestCafe versions to help us debug it quickly. If you cannot share a project, please try the following steps:
 - [Install a previous Chromium version](https://superuser.com/questions/920523/where-can-i-download-old-stable-builds-of-chromium-from) and check if the issue can be reproduced with it
 - Install a different TestCafe version (1.8.7, 1.8.6, etc.)
 - Provide a recorded video
 - Decrease concurrency level

**Comment by fissshka** [comment]
I've identified the root cause in my case. 
If CPU usage on my machine exceeds 70-80% this error occurs. 
In case of lower CPU usage TestCafe works as expected.

**Comment by tklepacki** [comment]
Hi. Same for me:

Teamcity Buildlog:

[16:15:14][Step 6/7] GeneralError: Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure.
[16:15:14][Step 6/7]     at BrowserSet._waitConnectionsOpened (C:\BuildAgent\work\17956cc26defe6f4\src\Signal.FrontendTesting\node_modules\testcafe\src\runner\browser-set.ts:91:30)
[16:15:14][Step 6/7]     at C:\BuildAgent\work\17956cc26defe6f4\src\Signal.FrontendTesting\node_modules\testcafe\src\runner\browser-set.ts:114:35
[16:15:14][Step 6/7]     at processTicksAndRejections (internal/process/task_queues.js:97:5)
[16:15:14][Step 6/7]     at Bootstrapper._getBrowserConnections (C:\BuildAgent\work\17956cc26defe6f4\src\Signal.FrontendTesting\node_modules\testcafe\src\runner\bootstrapper.ts:215:16)
[16:15:14][Step 6/7]     at async Promise.all (index 0)
[16:15:14][Step 6/7]     at Bootstrapper._bootstrapParallel (C:\BuildAgent\work\17956cc26defe6f4\src\Signal.FrontendTesting\node_modules\testcafe\src\runner\bootstrapper.ts:391:38)
[16:15:14][Step 6/7]     at Bootstrapper.createRunnableConfiguration (C:\BuildAgent\work\17956cc26defe6f4\src\Signal.FrontendTesting\node_modules\testcafe\src\runner\bootstrapper.ts:424:42) {
[16:15:14][Step 6/7]   code: 'E1004',
[16:15:14][Step 6/7]   data: []
[16:15:14][Step 6/7] }
[16:15:14][Step 6/7] info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
[16:15:14][Step 6/7] error Command failed with exit code 1.

What is the web application and TestCafe test code?
Parameters:

target browser: chrome:headless
concurrency level: 3
reporter: teamcity
skipJsErrors: false
skipUncaughtErrors: false

TestCafe runner code:
const createTestCafe = require("testcafe");

createTestCafe().then(tc => {
  tc.createRunner()
    .src([
      'app/tests/cargoes/*.visual.ts',
      'app/tests/howToVideos/*.visual.ts',
      'app/tests/manualTonnageList/*.visual.ts',
      'app/tests/vessels/*.visual.ts',
    ])
    .concurrency(3)
    .reporter(['teamcity',
      {
        name: 'json',
        output: 'app/scripts/artifacts/reports/visual.json'
      },
      {
        name: 'xunit',
        output: 'app/scripts/artifacts/reports/visual.xml'
      }
    ])
    .run({
      quarantineMode: true
    }).then((failedCount) => {
      tc.close()
      process.exit(failedCount > 0 ? 1 : 0)
    })
    .catch((error) => {
      console.error(error)
      tc.close()
      process.exit(1)
    })
})

Custom configuration file:
{
    "browsers": "chrome:headless",
    "src": [
        "app/tests/"
    ],
    "screenshots": {
        "path": "app/scripts/artifacts/screenshots",
        "takeOnFails": true,
        
        "pathPattern": "${DATE}_${TIME}/${RUN_ID}/${TEST_ID}.png"
    },
    "quarantineMode": false,
    "concurrency": 1
}

Environment details:
testcafe version: 1.8.7
node.js version: v12.18.1
browser name and version: 83.0.4103.116 (Official Build) (64-bit)
platform and version: Microsoft Windows Server 2016 Datacenter
TeamCity: TeamCity Enterprise 2019.1 (build 65998)
testcafe-reporter-teamcity: 1.0.10

It is not seems to be a CPU usage problem but more random issue. Here is the machine performance graph where the issue didn't occur:

![image](https://user-images.githubusercontent.com/12895225/86137774-f04fd280-baed-11ea-9cbe-23a0247863d7.png)

70%-100% CPU usage - no issue in this particular run.

**Comment by AndreyBelym** [comment]
Thank you for the provided information. While your tests can pass sometimes even when CPU, RAM or disk is overloaded, the stability of your test system is very unpredictable. You can try to use the minimal concurrency level (1) and see if it helps.

**Comment by doverradio** [comment]
One way I was able to overcome this issue (maybe this will help somebody
else) is to set up Express on one server (Server A) and have additional
servers (Server B, Server C, etc.) which receive requests from server A
which, after receiving those requests, Server B/Server C perform the tests
in isolation and respond back to my Mongo Atlas cloud database (directly
from the test, in .afterEach), which then server A reads, bypassing the
drama of crashing the Server A app due to maxing out CPU/RAM.

**Comment by alexey-lin** [comment]
Hi @doverradio, Thank you for sharing your experience.

**Comment by tklepacki** [comment]
> Thank you for the provided information. While your tests can pass sometimes even when CPU, RAM or disk is overloaded, the stability of your test system is very unpredictable. You can try to use the minimal concurrency level (1) and see if it helps.

It is not a problem of CPU or RAM in my case. The issue happens also when the machine is not overloaded. Also I don't want to execute tests with the minimal concurrency because I would like my test to be run in reasonable times. Even if it worked It's just a workaround not the real answer to the problem.

**Comment by wentwrong** [comment]
Hello, @tklepacki,

This error indicates that a browser doesn't send a HTTP request to the TestCafe server. If it happens sporadically, the likely cause is the lack of resources (CPU/memory). When the performance is low, the connection cannot be established during the timeout specified by the TestCafe (2 minutes for local browsers and 6 minutes for remote or cloud browsers). 

If the issue you faced is different, please send us a minimal reproducible example.

**Comment by doverradio** [comment]
Hey guys, back.  Earlier I had posted that I overcame this issue by setting up additional servers.  While that solution worked for a few days, now all 5 of my Ubuntu testing servers are failing.  Each one of them previously worked without issue and then, due to a security issue I was forced to reinstall the entire OS for each server.  After, none of them work with TestCafe!  It's unbelievable.  I have detailed this in [issue 5314](https://github.com/DevExpress/testcafe/issues/5314).

**Comment by Farfurix** [comment]
@doverradio

Hello, I updated the #5314 thread.

**Comment by bryg217** [comment]
Hello,

@wentwrong, your response makes a ton of a sense!

However, since this issue is happening more frequently for us (2 or more times a week), **is there some sort of "retry" option we could use if the first connection attempt fails**?  BTW: We have `-q` (quarantine mode) enabled for our tests.  Does that retry the connection itself?

I think something like this could be useful b/c typically, we just manually re-run it via Jenkins, and then the test(s) run successfully.

Thanks!

**Comment by wentwrong** [comment]
Hello, @bryg217,

Quarantine mode doesn't retry the connection when it's not being established during the timeout. This mode is designed for retrying unstable tests - it'll not be helpful with an unstable browser connection.

We changed the error message so that it provides more information in the context of the following PR: [Improve the message shown by "Unable to establish connection" error](https://github.com/DevExpress/testcafe/pull/5720). Also, in this PR we introduced the functionality to specify the timeout during which browser connection should be established. The changes from the PR are not yet released/reflected in our documentation, but you can already test them and see if increasing the default timeout by specifying "--browser-init-timeout" flag resolves the problem with stability. You can install it by running the following command:
```
npm install https://git.io/JLQ2k
```

**Comment by bryg217** [comment]
Hello @wentwrong ,

Perfect!  That also makes sense and is **super helpful**.

Thank you for the quick turnaround on this, as well as being thorough.

My team and I look forward to using the updates once they are ready.

Thanks.

**Comment by wentwrong** [comment]
Hello @bryg217,

I'm glad this is helpful. Further feedback on whether the updates will solve the problem with stability would be much appreciated.

**Comment by Evilweed** [comment]
**I think we have found the reason.** 

@AndreyBelym 

We were moving TestCafe to run on Google Cloud Run, and we were using always only concurrency 1 (just one browser at a time). It was strange to us that even one browser was failing with this error.

We have tried:
- Closing browser after each run and cleaning env;
- Running TestCafe on separate process than our main runner;
- Killing the Google Cloud Run container after each test and using clean container;
- Update TestCafe and using browserInitTimeout;
- Creating new Dockerfile on Debian with newer chrome and better logging;

Nothing helped.

But... after bit of googling we have found that Puppeteer has similar issue. Googling for the solution for Puppeteer failed.

However, our Engineer started writing down everything that is different between good and bad starts, and he have found that ports were consistently same in failed runs.

So we have hardcoded the port like this:
```
return await createTestCafe('localhost',8081,8082)
```

**After hardcoding it like that - everything works perfectly.**

Maybe you could add some feature that checks if Port is available before trying to start process with this port?

**Comment by cdaringe** [comment]
We are finding this error _on the daily_ in chrome as well. We've gone nuts and added debug(...) statements everywhere :)

- _onIdle issues HTML to chrome
- 45-60s later, the browser closes

the HTML idle page document either doesn't make it to chrome (unlikely), or, the on page javascript is periodically failing, which prevents TC from bootstrapping itself.

I'd like to find a way to capture the local chrome output emitted localChrome.start(...). Anyone know if this is feasible?

**Comment by cdaringe** [comment]
Even with DEBUG=hammerhead*,testcafe* and with aggressively debug() enhanced source, we were not able to discern why chrome is not processing the IDLE_PAGE content. If any chrome wizards lurk in here, we attempted to use a "custom" testcafe browser, capture `chrome.log` logs & stdio from the foreground chrome process, but it didn't prove super useful because we were unable to get content from tabs/windows: https://stackoverflow.com/questions/66926607/how-can-i-run-chromium-in-the-foreground-and-capture-native-and-webview-logs . If we can reliably get tab level, network level, and window level events from the chromium cli & emitted logs, such information could perhaps improve robustness in the TC workflow.

The TestCafe initialization process can improve robustness, now, by changing the synchronization mechanisms executed between the browser process & node runner.

Current process:

- TC launches chrome, pointing to a page
- Chrome loads page
- Page does I/O w/ TC server
- Page processes init scripts, reporting back to server on each completion via POST
- POST handler allows test execution to resume

A more robust approach would be:

- TC launches chrome
- TC immediate gets CDP handle to chrome
- TC injects serialized init scripts into chrome, using request/response control flow

Currently, various effects are executed, and success is achieved by optimistically expecting that each downstream uncontrolled effect executes successfully. It is dangerous to pass ownership of initialization control flow to chrome, and chrome to the embedded webapp, as TC does not have hooks into _either_ of these systems by the time `runInitScripts` is called. TC launches chrome, passes a URL, &  🤞 wishes both chrome and the downstream web-app the best of luck.  What if TC managed the whole init process, vs implicitly marshaling that responsibility to these other (generally reliable) entities?

Current process (pseudo):
```
let testCafeWindowIsReadyDeferred = ... <actually implemented via initScriptsQueue>
TC (system 1) creates readiness deferred
TC launches Chrome (system 2)
Chrome tries to load a web app (system 3) 
Web-app attempts to _do lots of work_ in the window, which hopefully eventually lands a POST call into system 1, which settles the deferred
READY
```

Possible future:

```
TC (system 1) launches chrome (system 2)
TC navigates chrome to idle page (resolve/reject)
TC injects script, gets response (resolve/reject)
...loop, until all init scripts settled
READY
```

**Comment by wentwrong** [comment]
@Evilweed,
> Maybe you could add some feature that checks if Port is available before trying to start process with this port?

TestCafe [checks](https://github.com/DevExpress/testcafe/blob/57aa1d303ffbe820e5711361ec5c3707899116ed/src/index.js#L28) that ports are free before starting the process. If the check fails for some reason, TestCafe shows the following error message:
```
ERROR The specified 1337 port is already in use by another program.
```
So, the issue may be caused by restrictions related to ports that can be used inside a container in Google Cloud Run.

@cdaringe,
Thanks for sharing your idea. However, changing the existing way of synchronization between our process and browser should require major changes and may not be applicable to browsers other than Chrome.

> I'd like to find a way to capture the local chrome output emitted localChrome.start(...). Anyone know if this is feasible?

To collect Chrome debug logs, you can try the approach with a temporary Chrome profile. It's described in this answer: https://github.com/DevExpress/testcafe/issues/3862#issuecomment-502710725.
